src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -440,7 +481,11 @@ public void DoWriteIfNeeded()
                 }, this);
 
                 Self._head = _lockedEnd.Block;
-                Self._head.Start = _lockedEnd.Index;
```

Comment:
Wow, ok it is because the `Write` callback is completing before the rest of `DoWriteIfNeeded` so `OnWriteCompleted` is called before `_head` is set to `_lockedEnd.Block`

![callstack](http://aoa.blob.core.windows.net/aspnet/callstack.png)